proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,371 @@
+> this be deferred until we start working on ray tracing later?
+
+In the root signature DSL, a root signature is made up of a list of "root
+elements". The in-memory datastructures are designed around this concept where
+the RootSignature class is a vector of variants.
+
+Example:
+
+```c++
+RootSignature[
+ "RootFlags(ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT),"
+ "CBV(b0, space=1),"
+ "StaticSampler(s1),"
+ "DescriptorTable(SRV(t0, numDescriptors=unbounded))"
+]
+```
+
+When parsed will produce a the equivalent of:
+
+```c++
+// placeholder: update this once detailed design complete
+parsedRootSignature = RootSignature{
+  RootFlags(ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT),
+  RootCBV(0, 1),
+  StaticSampler(1),
+  DescriptorTable({
+    SRV(0, unbounded)
+  })
+};
+```
+
+### Root Signatures in the LLVM IR 
+
+During frontend code generation an IR-based representation of the root signature
+is generated from the in-memory data structures stored in the AST. This is
+stored as metadata nodes, identified by named metadata. The metadata format
+itself is a straightforward transcription of the in-memory data structure - so
+it is a list of root elements.
+
+See [Metadata Schema](#metadata-schema) for details.
+
+> **Open Question**: what should the named metadata be?  There's options I
+> think...
+
+Example for same root signature as above:
+
+```llvm
+; placeholder - update this once detailed design complete
+!directx.rootsignatures = !{!2}
+!2 = !{ptr @main, !3 }
+!3 = !{ !4, !5, !6, !7 }
+!4 = !{ !"RootFlags", i32 1 }
+!5 = !{ !"RootCBV", i32 1, i32 0 }
+!6 = !{ !"StaticSampler", i32 1, i32 0, ... }
+!7 = !{ !"DescriptorTable", !8 }
+!8 = !{ !"SRV", i32 0, i32 -1 }
+```
+
+
+### Code Generation
+
+During backend code generation, the LLVM IR metadata representation of the root
+signature is converted data structures that represent the root signature that
+are more closely aligned to the final file format. For example, root parameters
+and static samplers can be intermingled in the previous formats, but are now
+separated into separate arrays at this point to aid in serializing.
+
+Example for same root signature as above:
+
+```c++
+// placeholder - update this once detailed design complete
+rootSignature = RootSignature(
+  ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT,
+  { // parameters
+    RootCBV(0, 1),
+    DescriptorTable({
+      SRV(0, unbounded)
+    })
+  },
+  { // static samplers
+    StaticSampler(1)
+  });
+```
+
+At this point, final validation is performed to ensure that the root signature
+itself is valid. One key validation here is to check that each register is only
+bound once in the root signature. Validation is performed this late to avoid
+requiring every frontend implementation to perform the same validation.
+
+> **Open Question** - although it seems important to perform the above
+> validation here for IR that didn't come directly from clang, maybe there'd
+> still be value in performing the same validation in Sema? Maybe it's ok to
```

Comment:
Checking that each register is bound only once in the root signature should be performed initially in Sema. This allows for early error reporting and ensures we have a correct AST. 

What cannot be done in Sema is checking that all used resources are bound into the root signature. This is because optimization may remove code related to resource usage

Reply:
I think it is important that it is checked here as well, so I think you're proposing we should perform the check in multiple places?  It'd be great if we could assume that the IR representation was correct, but I don't know if we can really do that.

Reply:
Yes. Unless we can make sure all frontends generate legal root signature metadata, we'll have to validate it in DirectX backend.